### PR TITLE
ATI marketing utils

### LINKED ISSUE #7385: [P2] Create util functions ATI marketing params
**Parent Issue** https://github.com/bbc/simorgh/issues/6609

This is a great candidate for TDD :)

**Is your feature request related to a problem? Please describe.**
Some page requests on assets rendered by Simorgh can be part of a campaign being run by the marketing team. We can tell if a page request originates from a marketing campaign when the url will contains some additional query string parameters.

Different campaign types will have contain different values. For a given campaign type we should match these values and return them, in the correct order.

**Example campaign url**
```
/persian?at_medium=affiliate&at_campaign=73&at_type=partner&at_identifier=whatsapp&at_format=Link&at_creation=wspersian&at_variant=Editorial
```

**Describe the solution you'd like**
Create a set of util functions for constructing ATI marketing campaign parameters based on a campaign type.

If a particular value is not present in the url the output for that value should be blank.

The output should follow this structure:

```
xto=A-B-C-D-E-F-G
```

_If a particular value is not present in the url the output for that value should be blank. e.g. if the value for `F` is missing the expected output would be:_

**Each individual value should be URL encoded, some of the values will be wrapped in square brackets after the URL encoding. Ensure we are not double encoding the url parameters**

```
xto=A-B-C-D-E--G
```

Expected outputs based on campaign types:

**Campaign type == `affiliate`**

```
A = the letters al
B = the value of the at_campaign field
C = the value of the at_type field, wrapped in square brackets
D = the value of the at_identifier field, wrapped in square brackets
E = the value of the at_format field, wrapped in square brackets
F = the value of the at_creation field, wrapped in square brackets
G = the value of the at_variant field, wrapped in square brackets
```

**Campaign type == `sl`**

```
A = the letters SEC
B = the value of the at_campaign field
C = the value of the at_platform field, wrapped in square brackets
D = the value of the at_creation field, wrapped in square brackets
E = the value of the at_variant field, wrapped in square brackets
F = if the value of the at_network field is search, then F = S, if the value of the at_network field is content then F = C
G = the value of the at_term field, wrapped in square brackets
```

**Campaign type == `email`**

```
A = based on the at_emailtype field... acquisition -> EREC, retention -> EPR, promotion -> ES
B = the value of the at_campaign field
C = the value of the at_creation field, wrapped in square brackets
D = the value of the at_send_date field
E = the value of the at_link field, wrapped in square brackets
F = the value of the at_recipient_id field followed by the @ symbol and the value of the at_recipient_list field
```

**Campaign type == `display`**

```
A = AD
B = the value of the at_campaign field
C = the value of the at_creation field, wrapped in square brackets
D = the value of the at_variant field, wrapped in square brackets
E = the value of the at_format field, wrapped in square brackets
F = no value
G = the value of the at_general_placement field, wrapped in square brackets
H = the value of the at_detail_placement field, wrapped in square brackets
```

**Campaign type starts with `custom` e.g. custom123**

```
A = CS followed by the part of the at_medium field following the word custom (e.g. for custom123, it becomes CS123)
B = the value of the at_campaign field
C = the value of the at_custom_1 field, wrapped in square brackets
D = the value of the at_custom_2 field, wrapped in square brackets
E = the value of the at_custom_3 field, wrapped in square brackets
F = the value of the at_custom_4 field, wrapped in square brackets
```

**Campaign type == `xtor`**

Take the value of the `xtor` QS e.g.`?xtor=12345678` and return the following

```
xto=12345678
```

Note the difference in the parameter name, the output should not include the `r`

**Describe alternatives you've considered**

**Testing notes**
[Tester to complete]

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

- [ ] This feature is expected to need manual testing.

**Additional context**

### PR DESCRIPTION
**Pull request by rhenshaw56**
Resolves #7385 

**Overall change:**
_Adds utility functions to support the various ATI marketing campaigns._

**Code changes:**

- _Adds analytics utility functions to support the various ATI marketing campaigns._
- _Added unit tests._

---

- [X] I have assigned myself to this PR and the corresponding issues
- [X] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [X] I have assigned this PR to the Simorgh project

**Testing:**

- [X] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
- [ ] This PR requires manual testing

### CODE REVIEW

File: src/app/lib/analyticsUtils/analytics.const.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 const SUPPORTED_MEDIUM_CAMPAIGN_TYPES = [
   'affiliate',
-  's1',
+  'sl',
```
Comment (rhenshaw56) [inline src/app/lib/analyticsUtils/analytics.const.js:3]: change made due to [value from doc](https://developers.atinternet-solutions.com/javascript-en/campaigns-javascript-en/marketing-campaigns-v2/#sponsored-links_6)

File: src/app/lib/analyticsUtils/index.js [hunk 2]
```diff
@@ -3,6 +3,7 @@ import uuid from 'uuid/v4';
 import pathOr from 'ramda/src/pathOr';
 import path from 'ramda/src/path';
 import Url from 'url-parse';
+
```
Comment (chris-hinds) [inline src/app/lib/analyticsUtils/index.js]:
```suggestion
```

File: src/app/lib/analyticsUtils/index.js [hunk 3]
```diff
@@ -333,4 +334,269 @@ export const getCampaignType = () => {
...
+const buildMarketingString = marketingValues =>
+  marketingValues
+    .map(({ value, wrap }) => (wrap && value ? `[${value}]` : value))
+    .join('-');
+
+export const getAffiliateMarketingString = href => {
+  const { query } = new Url(href, true);
+  return buildMarketingString([
+    {
+      description: ' affiliate campaign prefix',
```
Comment (chris-hinds) [inline src/app/lib/analyticsUtils/index.js]:
```suggestion
      description: 'affiliate campaign prefix',
```

File: src/app/lib/analyticsUtils/index.js [hunk 4]
```diff
@@ -333,4 +334,269 @@ export const getCampaignType = () => {
...
+    },
+    {
+      description: 'at_custom_4 field',
+      value: getMarketingUrlParam(query, 'at_custom4'),
+      wrap: true,
+    },
+  ]);
+};
+
+const parameteriseHash = hash => new Url(hash.replace('#', '?'), true).query;
```
Comment (chris-hinds) [inline src/app/lib/analyticsUtils/index.js]: Could we have a small comment explaining the use case for this function please.

File: src/app/lib/analyticsUtils/analytics.const.js (hunk 1 above)
Comment (andrewscfc) [reply src/app/lib/analyticsUtils/analytics.const.js:3]: Does this change need to be made back in the original issue?

File: src/app/lib/analyticsUtils/index.test.js [hunk 5]
```diff
@@ -606,3 +614,188 @@ describe('getCampaignType', () => {
...
+        ${'the value of the "at_campaign" field'}                                     | ${'https://www.bbc.com/mundo?at_medium=sl&at_campaign=73'}         | ${'SEC-73-----'}
+        ${'with the value of the "at_platform" field, wrapped in square brackets'}    | ${'https://www.bbc.com/mundo?at_medium=sl&at_platform=google'}     | ${'SEC--[google]----'}
+        ${'the value of the "at_creation" field, wrapped in square brackets'}         | ${'https://www.bbc.com/mundo?at_medium=sl&at_creation=my_adgroup'} | ${'SEC---[my_adgroup]---'}
+        ${'the value of the "at_variant" field, wrapped in square brackets'}          | ${'https://www.bbc.com/mundo?at_medium=sl&at_variant=Editorial'}   | ${'SEC----[Editorial]--'}
+        ${'the value of the "at_network" field when "at_network" field is "search"'}  | ${'https://www.bbc.com/mundo?at_medium=sl&at_network=search'}      | ${'SEC-----F=S-'}
+        ${'the value of the "at_network" field when "at_network" field is "content"'} | ${'https://www.bbc.com/mundo?at_medium=sl&at_network=content'}     | ${'SEC-----F=C-'}
```
Comment (andrewscfc) [inline src/app/lib/analyticsUtils/index.test.js]: Is it worth writing a test for `at_network` being something other than `search` or `content`? I guess you would want to omit the parameter if that somehow happened?

File: src/app/lib/analyticsUtils/index.test.js [hunk 6]
```diff
@@ -606,3 +614,188 @@ describe('getCampaignType', () => {
...
+});
+
+describe('getEmailMarketingString', () => {
+  describe('should return the "SEC" prefix', () => {
+    describe('with optional params', () => {
+      it.each`
+        expectation                                                                                                       | href                                                                                                             | expectedValue
+        ${'the value of "at_emailtype" field when its value is "acquisition"'}                                            | ${'https://www.bbc.com/mundo?at_medium=email&at_emailtype=acquisition'}                                          | ${'EREC-----'}
+        ${'the value of "at_emailtype" field when its value is "retention"'}                                              | ${'https://www.bbc.com/mundo?at_medium=email&at_emailtype=retention'}                                            | ${'EPR-----'}
+        ${'the value of "at_emailtype" field when its value is "promotion"'}                                              | ${'https://www.bbc.com/mundo?at_medium=email&at_emailtype=promotion'}                                            | ${'ES-----'}
```
Comment (andrewscfc) [inline src/app/lib/analyticsUtils/index.test.js]: Same as above, could probably do with a test for an unexpected value in `at_emailtype`

File: src/app/lib/analyticsUtils/index.test.js [hunk 7]
```diff
@@ -606,3 +614,188 @@ describe('getCampaignType', () => {
...
+        ${'the value of the "at_creation" field, wrapped in square brackets'}                                             | ${'https://www.bbc.com/mundo?at_medium=email&at_emailtype=promotion&at_creation=wsmundo'}                        | ${'ES--[wsmundo]---'}
+        ${'the value of the at_send_date field'}                                                                          | ${'https://www.bbc.com/mundo?at_medium=email&at_emailtype=promotion&at_send_date=20190401'}                      | ${'ES---20190401--'}
+        ${'the value of the at_link field, wrapped in square brackets'}                                                   | ${'https://www.bbc.com/mundo?at_medium=email&at_emailtype=promotion&at_link=cta_button'}                         | ${'ES----[cta_button]-'}
+        ${'the value of the at_recipient_id field followed by the @ symbol and the value of the at_recipient_list field'} | ${'https://www.bbc.com/mundo?at_medium=email&at_emailtype=promotion&at_recipient_id=5633&at_recipient_list=200'} | ${'ES-----5633@200'}
```
Comment (andrewscfc) [inline src/app/lib/analyticsUtils/index.test.js]: Do we need a test for `at_recipient_id` without `at_recipient_list` and vice versa? I guess in that situation you'd probably omit the parameter entirely?

File: src/app/lib/analyticsUtils/index.test.js [hunk 8]
```diff
@@ -606,3 +614,188 @@ describe('getCampaignType', () => {
...
+    });
+  });
+});
+
+describe('getCustomMarketingString', () => {
+  describe('should return the "CS" prefix followed by the part of the at_medium field following the word custom', () => {
+    describe('with optional params', () => {
+      it.each`
+        expectation                                                          | href                                                                | expectedValue
+        ${'the value of the "at_campaign" field'}                            | ${'https://www.bbc.com/mundo?at_medium=custom123&at_campaign=56'}   | ${'CS123-56----'}
```
Comment (andrewscfc) [inline src/app/lib/analyticsUtils/index.test.js:749]: Maybe add a test to cater for when this field doesn't contain `custom`, once again maybe just omit the field?

File: src/app/lib/analyticsUtils/index.test.js [hunk 9]
```diff
@@ -606,3 +614,188 @@ describe('getCampaignType', () => {
     expect(campaignType).toEqual('XTOR');
   });
 });
+
```
Comment (andrewscfc) [inline src/app/lib/analyticsUtils/index.test.js:617]: Just took a look back at @jimjohnsonrollings original issue, I think these query strings need to also work with a `#`: 

> The first uses a query string `?` / anchor `#` string including `at_medium=`

https://github.com/bbc/simorgh/issues/6609

You may want to double check that with him.

**Comment by chris-hinds** [comment]
@rhenshaw56 great work on this.

Thoughts on extracting these new utils out to their own file? This util file is getting quite large now.

I would recommend a new small refactor PR for this though

File: src/app/lib/analyticsUtils/index.js [hunk 10]
```diff
@@ -333,4 +333,272 @@ export const getCampaignType = () => {
   return null;
 };
 
+const getMarketingUrlParam = (query, field) =>
+  Object.prototype.hasOwnProperty.call(query, field) ? query[field] : '';
+
+const buildMarketingString = marketingValues =>
+  marketingValues
+    .map(({ value, wrap }) => (wrap && value ? `[${value}]` : value))
```
Comment (HarveyPeachey) [inline src/app/lib/analyticsUtils/index.js:353]: Just a small question about wrapping some `medium` parameter values in square brackets. As I can see in the code some of the values in the key:value pairs will be wrapped, but the campaign url example has no square brackets:
```
/persian?at_medium=affiliate&at_campaign=73&at_type=partner&at_identifier=whatsapp&at_format=Link&at_creation=wspersian&at_variant=Editorial
```

Are we 100% sure they need wrapping in square brackets?

File: src/app/lib/analyticsUtils/index.test.js [hunk 11]
```diff
@@ -606,3 +614,188 @@ describe('getCampaignType', () => {
...
+        expect(getCustomMarketingString(href)).toEqual(expected);
+      });
+    });
+  });
+});
+
+describe('getXtorMarketingString', () => {
+  it.each`
+    expectation                                                               | href                                                      | expectedValue
+    ${'the value of the "xtor" field when it is a hash param from an anchor'} | ${'https://www.bbc.com/mundo/#at_medium=sl&xtor=AD-3030'} | ${'AD-3030'}
```
Comment (andrewscfc) [inline src/app/lib/analyticsUtils/index.test.js:776]: Is this meant to include `at_medium` _and_ `xtor`?

From the original issue https://github.com/bbc/simorgh/issues/6609, it suggests that `at_medium` can only be present when a query string is used not a `#`:

> N.B. If a query string is used, then other key pairs can be present before or after the `xtor` or `at_medium` parameter.

Maybe another one to clarify with @jimjohnsonrollings

File: src/app/lib/analyticsUtils/index.js [hunk 12]
```diff
@@ -333,4 +333,272 @@ export const getCampaignType = () => {
...
+  if (!campaignType) return null;
+
+  const supportedCampaignMappings = {
+    affiliate: () => getAffiliateMarketingString(href),
+    sl: () => getSLMarketingString(href),
+    email: () => getEmailMarketingString(href),
+    display: () => getDisplayMarketingString(href),
+    custom: () => getCustomMarketingString(href),
+    XTOR: () => getXtorMarketingString(href),
+  };
```
Comment (andrewscfc) [inline src/app/lib/analyticsUtils/index.js:578]: I don't know if you want to add a try-catch around these calls, there is a lot of specific logic for each kind of marketing string and there is a possibility an error could be thrown, I don't know what would happen if we did throw in here, we don't have error boundaries in our app so not quite sure. Maybe at least test throwing an error from here an see what happens!

File: src/app/lib/analyticsUtils/index.js [hunk 13]
```diff
@@ -333,4 +333,272 @@ export const getCampaignType = () => {
...
+      value: getMarketingUrlParam(query, 'at_custom3'),
+      wrap: true,
+    },
+    {
+      description: 'at_custom_4 field',
+      value: getMarketingUrlParam(query, 'at_custom4'),
+      wrap: true,
+    },
+  ]);
+};
```
Comment (andrewscfc) [inline src/app/lib/analyticsUtils/index.js]: I think it would be good to break this file up so the different marketing strings are in their own files? Maybe a quick refactor PR after this one?

**Review by andrewscfc — COMMENTED** [review]
Hey love the format of the tests, I've picked out lots of little edge cases you may want to check with @jimjohnsonrollings and some that I can imagine have the possibility to throw errors.

**Comment by rhenshaw56** [comment]
> @rhenshaw56 great work on this.
> 
> Thoughts on extracting these new utils out to their own file? This util file is getting quite large now.
> 
> I would recommend a new small refactor PR for this though

@chris-hinds I guess we could always move those ATI specific ones to [some place within the ati folder](https://github.com/bbc/simorgh/tree/latest/src/app/containers/ATIAnalytics)

File: src/app/lib/analyticsUtils/index.test.js (hunk 5 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.test.js]: This was stated in the [ati docs on the `at_network` field](https://developers.atinternet-solutions.com/javascript-en/campaigns-javascript-en/marketing-campaigns-v2/#sponsored-links_6),
```
Depending on the source being a paid search result (search) or an ad display on content networks (content)
```
@andrewscfc I figured that, the `at_network` will either be a `search`, `content` or empty value(in which case it wouldn't be specified `(-)`). Perhaps @jimjohnsonrollings might know more but I don't think it will have other types

File: src/app/lib/analyticsUtils/index.test.js (hunk 6 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.test.js]: similar to the comment above,

[the docs on `at_emailtype` specify that](https://developers.atinternet-solutions.com/javascript-en/campaigns-javascript-en/marketing-campaigns-v2/#email-marketing_7),
```
Email campaigns can be split in 3 distinct and exclusive types: acquisition, retention, promotion.
```

File: src/app/lib/analyticsUtils/index.js (hunk 12 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.js:578]: @andrewscfc [there's this bit](https://github.com/bbc/simorgh/pull/7429/files#diff-1c3985090b52f7b568ce9f3651bd02abR599-R601) where this is null checked so that doesn't happen eventually

File: src/app/lib/analyticsUtils/index.test.js (hunk 7 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.test.js]: wasn't specified in the original issue, but will need to clarify from @jimjohnsonrollings

File: src/app/lib/analyticsUtils/index.test.js (hunk 11 above)
Comment (jimjohnsonrollings) [reply src/app/lib/analyticsUtils/index.test.js:776]: I believe that either can follow a `#`

File: src/app/lib/analyticsUtils/index.test.js (hunk 8 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.test.js:749]: The original issue and [the docs](https://developers.atinternet-solutions.com/javascript-en/campaigns-javascript-en/marketing-campaigns-v2/#custom-marketing-sources_10) both had examples with `custom`, but will defer to  @jimjohnsonrollings's thoughts on this

File: src/app/lib/analyticsUtils/index.test.js (hunk 5 above)
Comment (jimjohnsonrollings) [reply src/app/lib/analyticsUtils/index.test.js]: I don't expect another value - are you suggesting we should test that an different value returns the same as it not being specified?

File: src/app/lib/analyticsUtils/index.test.js (hunk 7 above)
Comment (jimjohnsonrollings) [reply src/app/lib/analyticsUtils/index.test.js]: I think if one is missing/empty then omit that one?

File: src/app/lib/analyticsUtils/analytics.const.js (hunk 1 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/analytics.const.js:3]: nopes, it was stated as `sl` in the original issue

File: src/app/lib/analyticsUtils/index.test.js (hunk 11 above)
Comment (jimjohnsonrollings) [reply src/app/lib/analyticsUtils/index.test.js:776]: Oh right.  I don't think you would ever expect both types together.  But either one can follow `?` or `#`

File: src/app/lib/analyticsUtils/index.test.js (hunk 9 above)
Comment (jimjohnsonrollings) [reply src/app/lib/analyticsUtils/index.test.js:617]: Yes, both types should work with `#` or `?`

File: src/app/lib/analyticsUtils/index.test.js (hunk 11 above)
Comment (jimjohnsonrollings) [reply src/app/lib/analyticsUtils/index.test.js:776]: The line Andrew quotes was meant to suggest that you could have something like `?renderer_env=test&at_medium=...`

File: src/app/lib/analyticsUtils/index.js (hunk 10 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.js:353]: @HarveyPeachey in the original issue, it was stated that they be wrapped in square brackets. kind of like some of the values in the `xtor` url format ie. `xtor=AD-3030-[ad_version7]-[without_text]-[468]-[www.site.com]-[GT]-[top_page]`

File: src/app/lib/analyticsUtils/index.test.js (hunk 7 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.test.js]: @jimjohnsonrollings does the `@` symbol remain in place and in the same order if we don't have either or both ?

File: src/app/lib/analyticsUtils/index.js (hunk 10 above)
Comment (jimjohnsonrollings) [reply src/app/lib/analyticsUtils/index.js:353]: In the case of `at_medium`, the strings arrive without square brackets, and we need to add them

File: src/app/lib/analyticsUtils/index.js (hunk 13 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.js]: @andrewscfc I was thinking since this is ati specific, might be worth having most of this all namespaced somewhere within the [ati folder](https://github.com/bbc/simorgh/tree/latest/src/app/containers/ATIAnalytics) and not necessarily in a generic analytic folder but would welcome thoughts from others on this

File: src/app/lib/analyticsUtils/index.test.js (hunk 7 above)
Comment (jimjohnsonrollings) [reply src/app/lib/analyticsUtils/index.test.js]: Yes I think so

File: src/app/lib/analyticsUtils/index.test.js (hunk 5 above)
Comment (andrewscfc) [reply src/app/lib/analyticsUtils/index.test.js]: @jimjohnsonrollings 
yeah, it's just testing the page doesn't blow up if something else ever comes through. 

> are you suggesting we should test that an different value returns the same as it not being specified?

yeah, just it is tolerant of garbage data, we don't get a broken empty page if ati breaks!

File: src/app/lib/analyticsUtils/index.test.js (hunk 6 above)
Comment (andrewscfc) [reply src/app/lib/analyticsUtils/index.test.js]: See https://github.com/bbc/simorgh/pull/7429#discussion_r468006583

File: src/app/lib/analyticsUtils/index.test.js (hunk 8 above)
Comment (andrewscfc) [reply src/app/lib/analyticsUtils/index.test.js:749]: See https://github.com/bbc/simorgh/pull/7429#discussion_r468006583

File: src/app/lib/analyticsUtils/index.js (hunk 13 above)
Comment (andrewscfc) [reply src/app/lib/analyticsUtils/index.js]: Sounds good, just good to follow SOLID with this stuff 👍

File: src/app/lib/analyticsUtils/index.js (hunk 12 above)
Comment (andrewscfc) [reply src/app/lib/analyticsUtils/index.js:578]: I was thinking if we had any cases that throw errors, this bit made me suggest it: https://github.com/bbc/simorgh/pull/7429/files#diff-1c3985090b52f7b568ce9f3651bd02abR439-R443 

If the value wasn't `aquisition`, `retention` or `promotion` it would return `null`. That though would probably just be inserted into the request string as `null-`. There probably isn't anything that throws an error so should be okay.

File: src/app/lib/analyticsUtils/index.js (hunk 13 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.js]: I'll create an issue to cater for this refactor

File: src/app/lib/analyticsUtils/index.test.js (hunk 5 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.test.js]: fixed!

File: src/app/lib/analyticsUtils/index.test.js (hunk 6 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.test.js]: fixed!

File: src/app/lib/analyticsUtils/index.test.js (hunk 7 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.test.js]: added!

File: src/app/lib/analyticsUtils/index.test.js (hunk 9 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.test.js:617]: fixed! this should work now. I've added some tests as well

File: src/app/lib/analyticsUtils/index.test.js (hunk 11 above)
Comment (rhenshaw56) [reply src/app/lib/analyticsUtils/index.test.js:776]: removed!

**Review by andrewscfc — APPROVED** [review]
LGTM now, thanks for checking all the weird edge cases!

**Review by Simon2828 — APPROVED** [review]
LGTM 👍

**Comment by rhenshaw56** [comment]
merging as this can't be manually tested right now.
